### PR TITLE
Add lazy loading for mlflow flavors

### PR DESCRIPTION
**Pull request by chenmoneygithub**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Currently we are putting every flavor import in a huge try-except block, which stops at the first failure. Also our current approach imposes many restrictions for modules due to the skinny testing requirements.  

Technically this change should cause no change on users' side.

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/__init__.py [hunk 1]
```diff
@@ -50,76 +51,110 @@
...
-        "openai",
-        "sentence_transformers",
-        "johnsnowlabs",
-    ]
-except ImportError:
-    # We are conditional loading these commands since the skinny client does
-    # not support them due to the pandas and numpy dependencies of MLflow Models
-    pass
+# Lazily load mlflow flavors to avoid excessive dependencies.
+catboost = LazyLoader("mlflow.catboost", globals(), "mlflow.catboost")
```
Comment (harupy) [inline mlflow/__init__.py:55]: Curious how this affects code autocomplete on Databricks notebook.

File: tests/tracking/fluent/test_fluent_autolog.py [hunk 2]
```diff
@@ -99,6 +99,10 @@ def disable_new_import_hook_firing_if_module_already_exists(request):
             yield
 
 
+# Reference the method so that the mock does not conflict with lazy loading.
+mlflow.tensorflow.autolog
+
+
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply tests/tracking/fluent/test_fluent_autolog.py]: This works, changed to this way.

After looking into this file closely, i feel this odd conflict could be caused by the fixtures. I don't have much insight into it, Haru would you mind taking a look after we check in the PR? No time pressure tho.

File: mlflow/__init__.py (hunk 1 above)
Comment (harupy) [reply mlflow/__init__.py:55]: On Databricks, we run code during the Python REPL setup. That may touch `mlflow.tesnroflow.xxx`.

Comment (chenmoneygithub) [reply mlflow/__init__.py:55]: Good call, very likely.

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**